src/components/common/dialog.tsx

```diff
@@ -0,0 +1,910 @@
+                this.dialog.validation_state[this.path] = {
+                    cached_value: val,
+                    cached_result: v
+                };
+            } else {
+                const state = this.dialog.validation_state[this.path];
+                state.cached_value = val;
+                state.cached_result = v;
+            }
+            console.log("VALIDATE", this.path, v);
+        }
+        if (v) {
+            this.dialog.validation[this.path] = v;
+            this.dialog.validation_failed = true;
+            this.dialog.online_validation = true;
+            this.dialog.update();
+        }
+    }
+}
+
+interface DialogStateEvents {
+    changed(): void;
+}
+
+export class DialogState<V> extends EventEmitter<DialogStateEvents> {
+    values: V;
+    validation: Record<string, string | undefined> = { };
+    validation_state: Record<string, DialogValidationState> = { };
+
+    // internal
+    busy: boolean = false;
+    validation_failed: boolean = false;
+    online_validation: boolean = false;
+
+    // for fields and buttons
+    actions_disabled: boolean = false;
+    cancel_disabled: boolean = false;
+
+    error: unknown = null;
+
+    /* eslint-disable no-use-before-define */
+    validate_callback: (dlg: DialogState<V>) => void;
+    /* eslint-enable */
+
+    // for DialogValue
+
+    update() {
+        this.actions_disabled = this.busy || this.validation_failed;
+        this.cancel_disabled = this.busy;
+        this.emit("changed");
+    }
+
+    trigger_validation(): void {
+        this.validation = { };
+        this.validation_failed = false;
+        this.validate_callback(this);
+        this.update();
+    }
+
+    rename_validation_state(path: string, from: number, to: number) {
+        const from_path = path + "." + String(from);
+        const to_path = path + "." + String(to);
+        if (from_path in this.validation_state) {
+            console.log("RENAME", from_path, to_path);
+            this.validation_state[to_path] = this.validation_state[from_path];
+            delete this.validation_state[from_path];
+        }
+        for (const k in this.validation_state) {
+            if (k.indexOf(from_path + ".") == 0) {
+                const to = to_path + k.substring(from_path.length);
+                console.log("RENAME.", k, to);
+                this.validation_state[to] = this.validation_state[k];
+                delete this.validation_state[k];
+            }
+        }
+    }
+
+    // For buttons
+
+    async run_action(func: (vals: V) => Promise<void>) {
+        this.error = null;
+        this.busy = true;
+        this.update();
+        if (!await this.validate()) {
+            this.busy = false;
+            this.update();
+            return;
+        }
+
+        try {
+            cockpit.assert(this.values);
+            await func(this.values);
+        } catch (ex) {
+            console.error(String(ex));
+            this.error = ex;
```

Comment:
These 3 added lines are not executed by any test.  [Details](https://logs-cockpit.apps.ocp.cloud.ci.centos.org/pull-2435-e22b04b7-20251204-081542-fedora-43-devel/Coverage/lcov/github-pr.diff.gcov.html)